content/en/blog/2022/2022-05-25-contextual-logging.md

```diff
@@ -0,0 +1,255 @@
+---
+layout: blog
+title: "Contextual Logging in Kubernetes 1.24"
+date: 2022-05-25
+slug: contextual-logging
+---
+
+ **Authors:** Patrick Ohly (Intel)
+
+The [Structured Logging Working
+Group](https://github.com/kubernetes/community/blob/master/wg-structured-logging/README.md)
+has added new capabilities to the logging infrastructure in Kubernetes
+1.24. This blog post explains how developers can take advantage of those to
+make log output more useful.
+
+## Structured Logging
+
+The goal for [structured
+logging](https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/1602-structured-logging/README.md)
+is to replace C-style formatting and the resulting opaque log strings with log
+entries that have a well-defined syntax for storing message and parameters
+separately, for example as a JSON struct.
+
+When using the traditional klog text output format for structured log calls,
+strings were originally printed with `\n` escape sequences, except when
+embedded inside a struct. For structs, log entries could still span multiple
+lines, with no clean way to split the log stream into individual entries:
+
+```
+I1112 14:06:35.783529  328441 structured_logging.go:51] "using InfoS" longData={Name:long Data:Multiple
+lines
+with quite a bit
+of text. internal:0}
+I1112 14:06:35.783549  328441 structured_logging.go:52] "using InfoS with\nthe message across multiple lines" int=1 stringData="long: Multiple\nlines\nwith quite a bit\nof text." str="another value"
+```
+
+Now `<` and `>` markers and indention are used to ensure that splitting at a
+klog header at the start of a line is reliable and the output remains readable
+for humans:
```

Comment:
so ... what happens if the message text contains a ">" ?

Reply:
It will be indented, in contrast to the `>` end marker which isn't.

Reply:
I thought they were indented differently - let me double check.

Reply:
Right, the value uses \t and the end marker a single space.

This isn't visible when looking at the diff, but renders okay when looking at the full file.

Reply:
ah, I see, so the delimiters are really "<\n" and "^>".

Probably does *not* need to be added to the blog post, then.

Reply:
Or to be very precise, `=<\n` and `^ >` (one space) with `^\t` for the lines in the middle :sweat_smile: 

Agreed, this doesn't need to be in the blog post.